Sources/SwiftFrameCore/Workers/TextRenderer.swift

```diff
@@ -0,0 +1,100 @@
+import AppKit
+import Foundation
+
+final class TextRenderer {
+
+    private let ky_MinFontSize: CGFloat = 1
+    private let ky_pointSizeTolerance: CGFloat = 1e-7
+
+    // MARK: - Frame Rendering
+
+    func render(text: String, font: NSFont, color: NSColor, alignment: TextAlignment, rect: NSRect, context: CGContext) throws {
+        let attributedString = try makeAttributedString(for: text, font: font, color: color, alignment: alignment)
+
+        context.saveGState()
+        defer { context.restoreGState() }
+
+        let frame = makeFrame(from: attributedString, in: rect)
+        CTFrameDraw(frame, context)
+    }
+
+    private func makeFrame(from attributedText: NSAttributedString, in rect: NSRect) -> CTFrame {
+        let path = CGPath(rect: rect, transform: nil)
+        let frameSetter = CTFramesetterCreateWithAttributedString(attributedText)
+        return CTFramesetterCreateFrame(frameSetter, CFRange(location: 0, length: attributedText.length), path, nil)
+    }
+
+    // MARK: - Fitting & Size Computations
+
+    /// Determines the maximum point size of the specified font that allows to render the given text onto a
+    /// particular number of lines
+    func maximumFontSizeThatFits(string: String, font: NSFont, alignment: TextAlignment, maxSize: CGFloat, size: CGSize) throws -> CGFloat {
+        guard !string.isEmpty else {
+            throw NSError(description: "Empty string was passed to TextRenderer")
+        }
+
+        let calculatedFontSize = try maxFontSizeThatFits(string: string, font: font, alignment: alignment, minSize: ky_MinFontSize, maxSize: maxSize, size: size)
+        // Subtract some small number to make absolutely sure text will be rendered completely
```

Comment:
Wow, odd that we have to do this. :(

Reply:
It's really really weird, one might assume that if `attributedString.boundingRect(...)` returns a size that fits our rect, that `CoreText` would be able to render that same string into the rect 🌚 but apparently not